### PR TITLE
(Start to) handle cli hyperlinks in the terminal used for pkg dev tasks

### PR DESCRIPTION
**Pull request by jennybc**
Addresses #5218

This PR makes suggested code hints like `testthat::snapshot_accept()` or `testthat::snapshot_review()` clickable and runnable from the integrated terminal where positron-r sends package development tasks, such as `devtools::test()`. "Runnable" in the sense that the code is run in the user's R console. This gets routed into the same handlers as clickable code entered directly in the console, so the existing safety guards are in place.

Screenshot 2024-11-18 at 1 04 51 PM

(Note that this PR does _not_ change behaviour for file hyperlinks, which I've crossed out in the screenshot above. Those hyperlinks currently get delegated to the operating system, so for many R users, a `.R` file will open in RStudio. Changing that is a separate problem #5409.)

### QA Notes

To see the new behaviour, you need a development version of the cli R package. I would install that via `pak::pak("r-lib/cli")`. (Context: https://github.com/r-lib/cli/pull/739.)

Then you need to run tests or `R CMD check` on a package with a failing snapshot test, in order to get a clickable invitation to accept a new snapshot or to review the proposed snapshot in a Shiny app. I have made a toy package that could be obtained via [`usethis::create_from_github("jennybc/clilinks")`](https://github.com/jennybc/clilinks). It has a couple of snapshot tests that will always fail 😄 because they attempt to snapshot a random number.

* Install dev cli: `pak::pak("r-lib/cli")`
* Identify a package that tickles this feature and open it in Positron, perhaps via: `usethis::create_from_github("jennybc/clilinks")`
* Use our gesture for running package tests: Ctrl/Cmd + Shift + T or select *R: Test R Package* from the command palette.
* Click on a "runnable code" hyperlink:
  - Note this requires Ctrl/Cmd click! This is a VS Code convention, not specific to us.
  - Note that you might have to click twice. This is a bug (or 2 bugs?) for which fixes are making their way through the system (https://github.com/microsoft/vscode/issues/230010). Again, not us.
  - You will need to grant the extension permission to handle such URIs, either as a one-off or more persistently. This is a VS Code feature.

Screenshot 2024-11-18 at 1 04 51 PM

Screenshot 2024-11-18 at 12 55 37 PM

If you click `testthat::snapshot_review()`, you should see the code execute in the R console and this Shiny app appears. Click on "Accept" (the fact that "Skip" seems broken is https://github.com/r-lib/testthat/issues/2025 and unrelated to this PR).

Screenshot 2024-11-18 at 12 55 47 PM

### CODE REVIEW

File: extensions/positron-r/package.json [hunk 1]
```diff
@@ -87,6 +87,12 @@
         "title": "%r.command.packageTest.title%",
         "shortTitle": "%r.menu.packageTest.title%"
       },
+      {
+        "command": "r.cliHyperlinks",
+        "category": "R",
+        "title": "%r.command.cliHyperlinks.title%",
+        "shortTitle": "%r.menu.cliHyperlinks.title%"
+      },
```
Comment (jennybc) [inline extensions/positron-r/package.json:95]: Will go away. Purely a development convenience, before I modified cli the way I needed.

File: extensions/positron-r/package.json [hunk 2]
```diff
@@ -654,5 +660,8 @@
     },
     "minimumRVersion": "4.2.0",
     "minimumRenvVersion": "1.0.9"
-  }
+  },
+  "enabledApiProposals": [
+    "terminalDataWriteEvent"
+  ]
```
Comment (jennybc) [inline extensions/positron-r/package.json:666]: Will go away. Purely a development convenience, when convincing myself that we could see OSC 8 hyperlinks.

File: extensions/positron-r/package.nls.json [hunk 3]
```diff
@@ -12,6 +12,7 @@
 	"r.command.packageBuild.title": "Build R Package",
 	"r.command.packageInstall.title": "Install R Package and Restart R",
 	"r.command.packageTest.title": "Test R Package",
+	"r.command.cliHyperlinks.title": "Emit cli hyperlinks",
```
Comment (jennybc) [inline extensions/positron-r/package.nls.json:15]: Will go away.

File: extensions/positron-r/package.nls.json [hunk 4]
```diff
@@ -26,6 +27,7 @@
 	"r.menu.packageBuild.title": "Build R Package",
 	"r.menu.packageInstall.title": "Install R Package and Restart R",
 	"r.menu.packageTest.title": "Test R Package",
+	"r.menu.cliHyperlinks.title": "Emit cli hyperlinks",
```
Comment (jennybc) [inline extensions/positron-r/package.nls.json:30]: Will go away.

File: extensions/positron-r/src/commands.ts [hunk 5]
```diff
@@ -118,6 +118,15 @@ export async function registerCommands(context: vscode.ExtensionContext) {
...
 
+		vscode.commands.registerCommand('r.cliHyperlinks', async () => {
+			const tasks = await getRPackageTasks();
+			const task = tasks.filter(task => task.definition.task === 'r.task.cliHyperlinks')[0];
+			const isInstalled = await checkInstalled(task.definition.pkg);
+			if (isInstalled) {
+				vscode.tasks.executeTask(task);
+			}
+		}),
+
```
Comment (jennybc) [inline extensions/positron-r/src/commands.ts:129]: Will go away.

File: extensions/positron-r/src/extension.ts [hunk 6]
```diff
@@ -43,4 +44,76 @@ export function activate(context: vscode.ExtensionContext) {
 			refreshTestExplorer(context);
 		}
 	});
+
+	// Listen for terminal output
+	const disposable = vscode.window.onDidWriteTerminalData(event => {
+		processTerminalOutput(event.data);
+	});
+	context.subscriptions.push(disposable);
```
Comment (jennybc) [inline extensions/positron-r/src/extension.ts:52]: Will go away. Purely a development convenience, when convincing myself that we could see OSC 8 hyperlinks.

File: extensions/positron-r/src/extension.ts [hunk 7]
```diff
@@ -43,4 +44,76 @@ export function activate(context: vscode.ExtensionContext) {
...
+		authority: uri.authority,
+		path: uri.path,
+		query: uri.query,
+		queryParams: queryParamsObject,
+		fragment: uri.fragment,
+		fsPath: uri.fsPath
+	};
+
+	const uriDetailsJson = JSON.stringify(uriDetails, null, 2);
+	vscode.window.showInformationMessage(`URI Details:\n${uriDetailsJson}`);
```
Comment (jennybc) [inline extensions/positron-r/src/extension.ts:97]: Will go away. A development convenience.

File: extensions/positron-r/src/tasks.ts [hunk 8]
```diff
@@ -53,7 +53,37 @@ export async function getRPackageTasks(editorFilePath?: string): Promise<vscode.
...
+				// eslint-disable-next-line @typescript-eslint/naming-convention
+				R_CLI_HYPERLINK_RUN_URL_FORMAT: 'positron://positron.positron-r/cli?command=x-r-run:{code}',
+				// eslint-disable-next-line @typescript-eslint/naming-convention
+				R_CLI_HYPERLINK_HELP: "TRUE",
+				// eslint-disable-next-line @typescript-eslint/naming-convention
+				R_CLI_HYPERLINK_VIGNETTE: "TRUE"
+			}
+		},
+		{
+			task: 'r.task.cliHyperlinks',
```
Comment (jennybc) [inline extensions/positron-r/src/tasks.ts:71]: Will go away.

File: extensions/positron-r/package.json [hunk 9]
```diff
@@ -3,7 +3,7 @@
   "displayName": "%displayName%",
   "description": "%description%",
   "version": "0.0.2",
-  "publisher": "vscode",
+  "publisher": "positron",
```
[the remark this answers is not in the document]
Comment (jennybc) [reply extensions/positron-r/package.json:6]: From #5268, we're going to use "positron" as publisher here.

**Review by juliasilge — CHANGES_REQUESTED** [review]
This is working well for me! 🎉 

My main question is about the file hyperlinks. I understand that this PR is not meant to address those but it leaves us in a state where _both_ kinds of hyperlinks look like you should click them but clicking one kind results in something nice and clicking another kind results in something QUITE UNPLEASANT. Is it possible not to make file hyperlinks clickable, since we're not yet handling them? If not, can we put this new feature behind a feature flag until https://github.com/posit-dev/positron/issues/5409 is resolved? Right now we have landed a combo of "looks just like what I _should_ click" plus "both surprising and awful result".

**Comment by jennybc** [comment]
Unfortunately it's not (yet?) possible to have the run/help/vignette hyperlinks *but not have* file hyperlinks. At least, AFAIK, it's not possible and I'm pretty up to speed on this part of cli right now.

I feel like we can safely merge and I will tackle #5409 in the near term. This is behind an implicit feature flag of sorts, in that we only activate the hyperlinks in the presence of truly bleeding edge cli (version 3.6.3.9001). It's not just any dev version, but a dev version I bumped to specifically for this (https://github.com/r-lib/cli/commit/9d6e9db195d2bb1e7231d467dec2e0571f81f48d). I think the few people who might install this version of cli can probably cope for the couple weeks until file hyperlinks fully catch up.

Plus, as an experiment, I just changed the macOS file affiliation for `.R` from RStudio to Positron and this also "solves" the problem right away.

File: extensions/positron-r/src/uri-handler.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	// cli 3.6.3.9001 gained support for configuring the URL format for hyperlinks.
+	const cliPkg = await session.packageVersion('cli', '3.6.3.9001');
+	const cliSupportsHyperlinks = cliPkg?.compatible ?? false;
+
+	if (!cliSupportsHyperlinks) {
+		// eslint-disable-next-line @typescript-eslint/naming-convention
+		return { R_CLI_HYPERLINKS: 'FALSE' };
+	}
```
Comment (jennybc) [inline extensions/positron-r/src/uri-handler.ts:70]: Here's where we exit early, without activating hyperlinks in the integrated terminal, in the absence of bleeding edge cli.

**Review by juliasilge — DISMISSED** [review]
> This is behind an implicit feature flag of sorts, in that we only activate the hyperlinks in the presence of truly bleeding edge cli (version 3.6.3.9001). 

I am not a fan of this getting merged into main as is, but this is a fair point. Can you confirm with Gabor and take on making sure that we do not get a surprise cli release before the file hyperlinks are fixed up?

Alternatively, you could [throw a quick `IsDevelopmentContext` around this](https://connect.posit.it/positron-wiki/feature-flagging.html#development-builds-only) to gate it from being exposed to users as is?

**Comment by jennybc** [comment]
Yes @gaborcsardi has confirmed there will be no surprise cli release.

I don't love the idea of the "dev build only" gate, because I am hoping to use this in my package dev work. If if we think it needs to be gated, I could add a setting for it.

**Comment by jennybc** [comment]
Feature flag is in.

**Review by juliasilge — APPROVED** [review]

**Pull request merged**